### ISSUE TITLE
Hi, a question on svelte widgets

### ISSUE THREAD

**Issue by tystol** [kran6a/kran6a#1]
Hello, sorry for opening a random issue, I wasn't sure how else to reach out and didn't want to add noise to the issue thread over here:
https://github.com/sveltejs/kit/issues/3882#issuecomment-2334983397

re:
> I build svelte-based widgets. Currently they are tiny (~20kb - 100kb) self-contained svelte apps. They include their own css, fonts, images, svgs and js and are then bundled to a single js file using an internal tool I hacked together. Those widgets are then stored on a database along with their configuration/theming metadata so they are available for the end users.

I'm at the early stages of looking at doing something similar. Have you considered open sourcing the tool you use to build the svelte widgets? Or can you point me in the direction of what you did to acheive this?

**Comment by kran6a** [comment]
Sadly I can't share the source code but I can point you in the right direction.
The first thing you need is to have an entrypoint that looks like this:

```ts
import {mount} from"svelte";
import App from"./src/App.svelte";
export default (target, props)=>mount(App, {target,, props})
```

This means you can use it from the parent app like
```ts
import Widget from 'MyWidget';
Widget(document.body, {foo: "bar"}); //Mount the widget in the body, change document.body to any HTMLElement you want to mount the widget in
```

If you want to generate a single-file widget (meaning CSS, images, svgs and JS are bundled together in a single .js file) you will need to tweak the vite.config.ts file
```ts
export default defineConfig({
    build: {
        rollupOptions: {
            output: {
                minifyInternalExports: true,
                format: 'iife',
                hoistTransitiveImports: true,
                preserveModules: false,
                entryFileNames: 'index.js', //This is the output file name
                esModule: true
            },
            treeshake: 'smallest',
        },
        lib: {
            formats: ['iife'],
            name: 'simple',
            entry: './app.ts' //This points to an entrypoint like the one defined in the previous code
        },
        target: 'esnext', //Optional, if you don't transpile you save bytes off the final bundle
        emptyOutDir: true,
        outDir: './dist',
        minify: true,
        cssMinify: 'lightningcss',
        assetsInlineLimit: 999999999, //Basically inline every asset
        reportCompressedSize: true
    },
    plugins: [
        svelte({
            emitCss: true, //This is required so that svelte CSS can be processed by other vite plugins
            preprocess: vitePreprocess(),
            compilerOptions: {
                discloseVersion: false,
                css: 'external',
                customElement: false,
                preserveWhitespace: false
            },
        }),
        cssInjectedByJsPlugin({ //imported from vite-plugin-css-injected-by-js
            topExecutionPriority: true //Append the css to document.head before writing to document.body
        }),
        inliner({entry: "./app.ts"}) //Custom plugin, does some optimizations
    ]
});
```

Assets referenced from svelte files can be automatically inlined by vite
```ts
    import click from 'assets/audio/click.mp3?inline'; //The mp3 file is base64 encoded by vite and added to the bundle
```

If you use stylesheets you must import them from within your app so they are bundled.
Your fonts should also be local and referenced in the css by their path.
This is an example taken from my code
```postcss
@font-face {
        font-family: 'Inter';
        font-style: normal;
        font-weight: 400 900;
        font-display: swap;
        src: url("./assets/fonts/inter_latin.woff2") format("woff2");
        unicode-range: U+21,U+23-26,U+28,U+29,U+2B-39,U+3D,U+3F-5A,U+5C,U+5F,U+61-7A,U+7C,U+A1,U+AA,U+BA,U+BF-C1,U+C7-C9,U+CC,U+CD,U+CF,U+D1-D3,U+D9,U+DA,U+DC,U+E0,U+E1,U+E7-E9,U+EC,U+ED,U+EF,U+F1-F3,U+F9,U+FA,U+FC,U+20AC;
    }
```
You can ignore the unicode-range. It does nothing interesting by itself other than restricting the characters the font will be applied to (this won't make your bundle smaller).

I think this is the most basic single-file setup I can think of.
If you want to dive deeper you will need custom-made tools. I can name a few so you can look them up and think whether they are worth the effort or not:

1.  Font subsetter. Fonts are very big assets because they include lots of glyphs (greek alphabet, russian alphabet, lot of symbols, ligatures, etc.), weights and variants. I use a custom plugin to strip anything outside the desired unicode-ranges and any non-used weight/variant from the source font. The plugin automatically reads the unicode-range, font-weight and font-style from the CSS code. Note that if you use a glyph that has been stripped it will either look like a broken character or it will use a default font.
2. SVG inliner. You should read [this article](https://codepen.io/tigt/post/optimizing-svgs-in-data-uris). By implementing this encoding instead of base64 you can save a few bytes per svg. Probably not worth the effort as it only applies to svgs that are used within CSS and should not be used on CSS used in HTML.
3. Service worker and workers, PWA. This was the most difficult thing I had to implement for single-file bundles as there is zero documentation online about it. Probably your widgets shouldn't register a service worker / worker but I had the need to make single file SPAs. It is indeed possible to create a single-file PWA that is recognized as such by the browser (tested on Chromium and Firefox), including having its own manifest.json. I it can only be done if you render the "widget" within an iframe as I was not successfull in overriding the parent PWA with the "widget" PWA.

**Comment by tystol** [comment]
Thankyou - that is amazingly helpful!!

One question on the usage from the parent app:

> This means you can use it from the parent app like
> 
> ```ts
> import Widget from 'MyWidget';
> Widget(document.body, {foo: "bar"}); //Mount the widget in the body, change document.body to any HTMLElement you want to mount the widget in
> ```

In my case, I'll need to dynamically load the widget component from the server at runtime. Is that just a matter of using dynamic imports? eg. something like:

``
let Widget = await import(`./${widgetPath}.js`);
Widget(document.body, {foo: "bar"});
``

**Comment by kran6a** [comment]
Yes, dynamic imports would work, after all it is a vanilla js file exporting a default function that mounts the widget in a DOM node.

**Comment by tystol** [comment]
thanks mate, really appreciate this! I'll have a play and let you know how it goes.
